### PR TITLE
Fix a couple of issues with dotnet-watch

### PR DESCRIPTION
**Pull request by pranavkm**
* Print compilation errors in red and don't say hot reload succeeded when there's errors
* Fix an issue where the browser does not refresh after changes
* Fix an issue where deltas are not applied on assemblies loaded after the fact.
* Fix an issue where hot reload no longer works after an app restart.
* Fix a local dev issue to avoid using the local SDK when running dotnet-watch as an executable

### CODE REVIEW

File: src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs [hunk 1]
```diff
@@ -196,6 +196,10 @@ public void ApplyDeltas(IReadOnlyList<UpdateDelta> deltas)
                     {
                         System.Reflection.Metadata.AssemblyExtensions.ApplyUpdate(assembly, item.MetadataDelta, item.ILDelta, ReadOnlySpan<byte>.Empty);
                     }
+
+                    // Additionally stash the deltas away so it may be applied to assemblies loaded later.
+                    var cachedDeltas = _deltas.GetOrAdd(item.ModuleId, static _ => new());
```
Comment (pranavkm) [inline src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs:201]: We were applying these deltas, but I missed storing them.

File: src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs [hunk 2]
```diff
@@ -53,11 +53,6 @@ public static async Task ReceiveDeltas(HotReloadAgent hotReloadAgent)
             Log("Attempting to apply deltas.");
 
             hotReloadAgent.ApplyDeltas(update.Deltas);
-
-            // We want to base this off of mvids, but we'll figure that out eventually.
-            var applyResult = update.ChangedFile is string changedFile && changedFile.EndsWith(".razor", StringComparison.Ordinal) ?
```
Comment (pranavkm) [inline src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs:58]: I changed the heuristic about when we refresh the browser in web apps - we won't it if it looks like a Blazor app (presence of any .razor file). Works better than relying on individual files.

File: src/BuiltInTools/dotnet-watch/HotReload/CompilationHandler.cs [hunk 3]
```diff
@@ -171,7 +179,7 @@ private ImmutableArray<string> GetDiagnostics(Solution solution, CancellationTok
                     if (item.Severity == DiagnosticSeverity.Error)
                     {
                         var diagnostic = CSharpDiagnosticFormatter.Instance.Format(item);
-                        _reporter.Output(diagnostic);
+                        _reporter.Output("\x1B[40m\x1B[31m" + diagnostic);
```
Comment (pranavkm) [inline src/BuiltInTools/dotnet-watch/HotReload/CompilationHandler.cs:182]: Pretty red color

File: src/BuiltInTools/dotnet-watch/Program.cs [hunk 4]
```diff
@@ -67,7 +67,13 @@ public Program(IConsole console, string workingDirectory)
             // AppContext.BaseDirectory = $sdkRoot\$sdkVersion\DotnetTools\dotnet-watch\$version\tools\net6.0\any\
             // MSBuild.dll is located at $sdkRoot\$sdkVersion\MSBuild.dll
             var sdkRootDirectory = Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "..", "..");
+#if DEBUG
+            // In the usual case, use the SDK that contains the dotnet-watch. However during local testing, it's
+            // much more common to run dotnet-watch from a different SDK. Use the ambient SDK in that case.
+            MSBuildLocator.RegisterDefaults();
```
Comment (pranavkm) [inline src/BuiltInTools/dotnet-watch/Program.cs:73]: Annoying local build issue where it would use the dogfood SDK. This was causing the annoying FailFast error that I poked you about.

File: src/BuiltInTools/dotnet-watch/HotReload/DefaultDeltaApplier.cs [hunk 5]
```diff
@@ -29,14 +30,8 @@ public DefaultDeltaApplier(IReporter reporter)
 
         public bool SuppressBrowserRefreshAfterApply { get; init; }
 
-        public async ValueTask InitializeAsync(DotNetWatchContext context, CancellationToken cancellationToken)
+        public ValueTask InitializeAsync(DotNetWatchContext context, CancellationToken cancellationToken)
         {
-            if (_pipe is not null)
```
Comment (captainsafia) [inline src/BuiltInTools/dotnet-watch/HotReload/DefaultDeltaApplier.cs:34]: Why was this removed? It seems like good hygiene to have it?

**Review by captainsafia — APPROVED** [review]

File: src/BuiltInTools/dotnet-watch/HotReload/DefaultDeltaApplier.cs (hunk 5 above)
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/HotReload/DefaultDeltaApplier.cs:34]: Turns out we were newing up these instances in every hot reload loop. So this was never true. Also why the named pipe name is a static because we were only configuring it once.
